engine-types/src/types.rs

```diff
@@ -104,6 +108,114 @@ impl Mul<EthGas> for u64 {
     }
 }
 
+#[derive(
+    Default, Debug, Clone, Copy, Eq, PartialEq, Ord, PartialOrd, BorshSerialize, BorshDeserialize,
+)]
+/// Engine `fee` type which wraps an underlying u128.
+pub struct Fee(u128);
+
+impl Display for Fee {
+    fn fmt(&self, f: &mut Formatter<'_>) -> crate::fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl Fee {
+    /// Constructs a new `Fee` with a given u128 value.
+    pub const fn new(fee: u128) -> Fee {
+        Self(fee)
+    }
+
+    /// Consumes `Fee` and returns the underlying type.
+    pub fn into_u128(self) -> u128 {
+        self.0
+    }
+}
+
+impl Add<Fee> for Fee {
+    type Output = Fee;
+
+    fn add(self, rhs: Fee) -> Self::Output {
+        Fee(self.0 + rhs.0)
+    }
+}
+
+impl Add<Fee> for u128 {
+    type Output = Fee;
+
+    fn add(self, rhs: Fee) -> Self::Output {
+        Fee(self + rhs.0)
+    }
+}
+
+impl Add<u128> for Fee {
+    type Output = Fee;
+
+    fn add(self, rhs: u128) -> Self::Output {
+        Fee(self.0 + rhs)
+    }
+}
+
+impl Mul<Fee> for Fee {
+    type Output = Fee;
+
+    fn mul(self, rhs: Fee) -> Self::Output {
+        Fee(self.0 * rhs.0)
+    }
+}
+
+impl Mul<Fee> for u128 {
+    type Output = Fee;
+
+    fn mul(self, rhs: Fee) -> Self::Output {
+        Fee(self * rhs.0)
+    }
+}
+
+impl Mul<u128> for Fee {
+    type Output = Fee;
+
+    fn mul(self, rhs: u128) -> Self::Output {
+        Fee(self.0 * rhs)
+    }
+}
+
+impl Div<Fee> for Fee {
+    type Output = Fee;
+
+    fn div(self, rhs: Fee) -> Self::Output {
+        Fee(self.0 / rhs.0)
+    }
+}
+
+impl Div<Fee> for u128 {
+    type Output = Fee;
+
+    fn div(self, rhs: Fee) -> Self::Output {
+        Fee(self / rhs.0)
+    }
+}
+
+impl Div<u128> for Fee {
+    type Output = Fee;
+
+    fn div(self, rhs: u128) -> Self::Output {
+        Fee(self.0 / rhs)
+    }
+}
+
+impl From<u128> for Fee {
+    fn from(fee: u128) -> Self {
+        Self(fee)
+    }
+}
+
+impl From<Fee> for u128 {
+    fn from(fee: Fee) -> Self {
+        fee.0
+    }
+}
```

Comment:
Nit: This is less important than my other comments, but I also think implementation should not exist. All throughout our core logic fees are fees and so will never need to be unwrapped back to the raw numeric type. It is only at the boundary where we need to pass the value out of our application where we will need this conversion. This means it should not be needed very often and therefore having a super convenient way to do it is not needed.

On the other hand, convenience can erode the safety of new types because it allows the developer to change between types without needing to think about it, even if they _should_ be thinking about it.

Instead of `impl From<Fee> for u128` I prefer to have a function like

```rust
impl Fee {
  pub fn raw(self) -> u128 { self.0 }
}
```

The name "raw" is meant to suggest that you are dropping the semantics of the new type to work directly with the underlying numeric type. And again this should only happen at the boundaries of our application.

Reply:
Removed